# Самостоятельное задание 1 (+1 балл к оценке за курс).

## Колчина Анастасия БПМИ195 ##

In [2]:
import pandas as pd

#### Описание данных

В папке Data находится информация о студентах. Всего 10 групп студентов. Файлы делятся на две категории:
    * Students_info_i - информация о студентах из группы i
    * Students_marks_i - оценки студентов из группы i за экзамены

### Одно из важных достоинств pandas $-$ это удобные методы реляционного взаимодействия с данными, аналогичные, например, возможностям SQL для слияния и конкатенации таблиц: merge, join, concat. Наличие готовых методов позволяет не реализовывать самостоятельно поэлементную обработку данных и оперировать сразу целыми таблицами данных.

Подробно об этих методах посмотрите тут: https://www.kaggle.com/residentmario/renaming-and-combining#Combining

#### Задание 1. Соберите всю информацию о студентах в одну таблицу df. В получившейся таблице должна быть информация и оценки всех студентов из всех групп. Напечатайте несколько строк таблицы для демонстрации результата.¶

In [3]:
to_concat = []
for i in range(10):
    info = pd.read_csv("/Users/Anastasia/Documents/Activities-analysis-main/Data/Students_info_"+str(i)+".csv")
    marks = pd.read_csv("/Users/Anastasia/Documents/Activities-analysis-main/Data/Students_marks_"+str(i)+".csv")
    group = info.join(marks.set_index('index'), on='index')
    to_concat.append(group)
df = pd.concat(to_concat)
df = df.reset_index()
df = df.drop("level_0", axis=1)
df

index  gender race/ethnicity parental level of education         lunch  \
0        0  female        group B           bachelor's degree      standard   
1        1  female        group C                some college      standard   
2        2  female        group B             master's degree      standard   
3        3    male        group A          associate's degree  free/reduced   
4        4    male        group C                some college      standard   
..     ...     ...            ...                         ...           ...   
995    995  female        group E             master's degree      standard   
996    996    male        group C                 high school  free/reduced   
997    997  female        group C                 high school  free/reduced   
998    998  female        group D                some college      standard   
999    999  female        group D                some college  free/reduced   

    test preparation course    group  math score  reading score  writing score  
0                      none   group1          72             72             74  
1                 completed   group1          69             90             88  
2                      none   group1          90             95             93  
3                      none   group1          47             57             44  
4                      none   group1          76             78             75  
..                      ...      ...         ...            ...            ...  
995               completed  group10          88             99             95  
996                    none  group10          62             55             55  
997               completed  group10          59             71             65  
998               completed  group10          68             78             77  
999                    none  group10          77             86             86  

[1000 rows x 10 columns]

#### Задание 2. Удалите столбец index у полученной таблицы. Напечатайте первые 10 строк таблицы.

In [4]:
df = df.drop("index", axis=1)
df.head(10)

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   
6  female        group B                some college      standard   
7    male        group B                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test preparation course   group  math score  reading score  writing score  
0                    none  group1          72             72             74  
1               completed  group1          69             90             88  
2                    none  group1          90             95             93  
3                    none  group1          47             57             44  
4                    none  group1          76             78             75  
5                    none  group1          71             83             78  
6               completed  group1          88             95             92  
7                    none  group1          40             43             39  
8               completed  group1          64             64             67  
9                    none  group1          38             60             50

#### Задание 3. Выведите на экран размеры полученной таблицы

In [5]:
df.shape

(1000, 9)

#### Задание 4. Выведите на экран статистические характеристики числовых столбцов таблицы (минимум, максимум, среднее значение, стандартное отклонение)

In [6]:
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

#### Задание 5. Проверьте, есть ли в таблице пропущенные значения

В таблице нет пропусков, т.к. число строк равно 1000, а count по всем трем предметам также равен 1000.

#### Задание 6. Выведите на экран средние баллы студентов по каждому предмету (math, reading, writing)

In [7]:
df.mean()

math score       66.089
reading score    69.169
writing score    68.054
dtype: float64

**Задание 7. Как зависят оценки от того, проходил ли студент курс для подготовки к сдаче экзамена (test preparation course)? Выведите на экран для каждого предмета в отдельности средний балл студентов, проходивших курс для подготовки к экзамену и не проходивших курс.**

In [8]:
df[df['test preparation course'] == 'completed'].mean()

math score       69.695531
reading score    73.893855
writing score    74.418994
dtype: float64

In [9]:
df[df['test preparation course'] == 'none'].mean()

math score       64.077882
reading score    66.534268
writing score    64.504673
dtype: float64

У студентов, не проходивших курс для подготовки, средний балл ниже, чем у подготовленных студентов.

**Задание 8. Выведите на экран все различные значения из столбца lunch.**

In [10]:
set(df['lunch'])

{'free/reduced', 'standard'}

**Задание 9. Переименуйте колонку "parental level of education" в "education", а "test preparation course" в "test preparation" с помощью метода pandas rename**
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html

In [11]:
df = df.rename(columns={"parental level of education": "education", "test preparation course": "test preparation"})
df

gender race/ethnicity           education         lunch test preparation  \
0    female        group B   bachelor's degree      standard             none   
1    female        group C        some college      standard        completed   
2    female        group B     master's degree      standard             none   
3      male        group A  associate's degree  free/reduced             none   
4      male        group C        some college      standard             none   
..      ...            ...                 ...           ...              ...   
995  female        group E     master's degree      standard        completed   
996    male        group C         high school  free/reduced             none   
997  female        group C         high school  free/reduced        completed   
998  female        group D        some college      standard        completed   
999  female        group D        some college  free/reduced             none   

       group  math score  reading score  writing score  
0     group1          72             72             74  
1     group1          69             90             88  
2     group1          90             95             93  
3     group1          47             57             44  
4     group1          76             78             75  
..       ...         ...            ...            ...  
995  group10          88             99             95  
996  group10          62             55             55  
997  group10          59             71             65  
998  group10          68             78             77  
999  group10          77             86             86  

[1000 rows x 9 columns]

**Зафиксируем минимальный балл для сдачи экзамена**

In [12]:
passmark = 50

**Задание 10. Ответьте на вопросы:**
    * Какая доля студентов сдала экзамен по математике (passmark > 50)?
    * Какая доля студентов, проходивших курс подготовки к экзамену, сдала экзамен по математике?
    * Какая доля женщин, не проходивших курс подготовки к экзамену, не сдала экзамен по математике? 

Я не совсем поняла про минимальный балл, но будем считать, что экзамен сдан, если студент набрал 50 баллов или больше.

In [13]:
len(df[df['math score'] >= passmark]) / 1000 # т.к. 1000 студентов всего

0.865

In [14]:
len(df[(df['test preparation'] == 'completed') & (df['math score'] >= passmark)]['math score']) / len(df[df['test preparation'] == 'completed']['math score'])

0.9217877094972067

In [15]:
len(df[(df['test preparation'] == 'none') & (df['gender'] == 'female') & (df['math score'] < passmark)]['math score']) / len(df[(df['test preparation'] == 'none') & (df['gender'] == 'female')]['math score'])

0.20958083832335328

**Задание 11. С помощью groupby выполните задания ниже. Также выведите время выполнения каждого из заданий.**
    * Для каждой этнической группы выведите средний балл за экзамен по чтению
    * Для каждого уровня образования выведите минимальный балл за экзамен по письму

In [23]:
import time

start = time.time()

print(df.groupby('race/ethnicity')['reading score'].mean())

end = time.time()
first_opt = end - start
print('Execution time (sec):', first_opt)

race/ethnicity
group A    64.674157
group B    67.352632
group C    69.103448
group D    70.030534
group E    73.028571
Name: reading score, dtype: float64
Execution time (sec): 0.0029990673065185547


In [24]:
start = time.time()

print(df.groupby('education')['writing score'].min())

end = time.time()
second_opt = end - start
print('Execution time (sec):', second_opt)

education
associate's degree    35
bachelor's degree     38
high school           15
master's degree       46
some college          19
some high school      10
Name: writing score, dtype: int64
Execution time (sec): 0.0029773712158203125


**Задание 12. Выполните задание 11 с помощью циклов. Сравните время выполнения.**

In [25]:
start = time.time()

points = {'group A': 0, 'group B': 0, 'group C': 0, 'group D': 0, 'group E': 0}
students = {'group A': 0, 'group B': 0, 'group C': 0, 'group D': 0, 'group E': 0}
for i in range(1000):
    row = df.loc[i,:]
    points[row['race/ethnicity']] += row['reading score']
    students[row['race/ethnicity']] += 1
for item in points:
    points[item] /= students[item]
print(points)

end = time.time()
first_dumb = end - start
print('Execution time (sec):', first_dumb)

{'group A': 64.67415730337079, 'group B': 67.35263157894737, 'group C': 69.10344827586206, 'group D': 70.03053435114504, 'group E': 73.02857142857142}
Execution time (sec): 0.16750812530517578


In [26]:
start = time.time()

min_edu = {"associate's degree": 100, "bachelor's degree": 100, "high school": 100, "master's degree": 100, "some college": 100, "some high school": 100}
for i in range(1000):
    row = df.loc[i, :]
    if (row['writing score'] < min_edu[row['education']]):
        min_edu[row['education']] = row['writing score']
print(min_edu)
        
end = time.time()
second_dumb = end - start
print('Execution time (sec):', second_dumb)

{"associate's degree": 35, "bachelor's degree": 38, 'high school': 15, "master's degree": 46, 'some college': 19, 'some high school': 10}
Execution time (sec): 0.13347768783569336


**Вывод:** лучше использовать встроенные пандасовские функции, т.к. они работают намного эффективнее и во много раз быстрее, чем простой перебор циклом.

Бонус: дабы не быть голословными, даже посчитаем, во сколько раз быстрее:

In [27]:
print("Первая подзадача быстрее решается через groupby, чем циклами в", first_dumb / first_opt, "раз")
print("Вторая подзадача быстрее решается через groupby, чем циклами в", second_dumb / second_opt, "раз")

Первая подзадача быстрее решается через groupby, чем циклами в 55.85340647110263 раз
Вторая подзадача быстрее решается через groupby, чем циклами в 44.83071748878924 раз


**Задание 13. Выведите на экран средние баллы студентов по каждому предмету в зависимости от пола и уровня образования. То есть должно получиться количество групп, равных 2 * (число уровней образования), и для каждой такой группы выыведите средний балл по каждому из предметов.**

Это можно сделать с помощью сводных таблиц (pivot_table):

https://www.kaggle.com/kamilpolak/tutorial-how-to-use-pivot-table-in-pandas

In [21]:
grouped = pd.pivot_table(df, index=['gender','education'])
grouped

math score  reading score  writing score
gender education                                                   
female associate's degree   65.250000      74.120690      74.000000
       bachelor's degree    68.349206      77.285714      78.380952
       high school          59.351064      68.202128      66.691489
       master's degree      66.500000      76.805556      77.638889
       some college         65.406780      73.550847      74.050847
       some high school     59.296703      69.109890      68.285714
male   associate's degree   70.764151      67.433962      65.405660
       bachelor's degree    70.581818      68.090909      67.654545
       high school          64.705882      61.480392      58.539216
       master's degree      74.826087      73.130435      72.608696
       some college         69.009259      64.990741      63.148148
       some high school     67.840909      64.693182      61.375000

#### Задание 14. Сколько студентов успешно сдали экзамен по математике?

Создайте новый столбец в таблице df под названием Math_PassStatus и запишите в него F, если студент не сдал экзамен по математике (балл за экзамен < passmark), и P иначе.

Посчитайте количество студентов, сдавших и не сдавших экзамен по математике.

Сделайте аналогичные шаги для экзаменов по чтению и письму.

In [37]:
df["Math_PassStatus"] = df["math score"].apply(lambda x: "F" if x < passmark else "P")
df["Math_PassStatus"].value_counts()

P    865
F    135
Name: Math_PassStatus, dtype: int64

In [38]:
df["Reading_PassStatus"] = df["reading score"].apply(lambda x: "F" if x < passmark else "P")
df["Reading_PassStatus"].value_counts()

P    910
F     90
Name: Reading_PassStatus, dtype: int64

In [39]:
df["Writing_PassStatus"] = df["writing score"].apply(lambda x: "F" if x < passmark else "P")
df["Writing_PassStatus"].value_counts()

P    886
F    114
Name: Writing_PassStatus, dtype: int64

In [40]:
df

gender race/ethnicity           education         lunch test preparation  \
0    female        group B   bachelor's degree      standard             none   
1    female        group C        some college      standard        completed   
2    female        group B     master's degree      standard             none   
3      male        group A  associate's degree  free/reduced             none   
4      male        group C        some college      standard             none   
..      ...            ...                 ...           ...              ...   
995  female        group E     master's degree      standard        completed   
996    male        group C         high school  free/reduced             none   
997  female        group C         high school  free/reduced        completed   
998  female        group D        some college      standard        completed   
999  female        group D        some college  free/reduced             none   

       group  math score  reading score  writing score Math_PassStatus  \
0     group1          72             72             74               P   
1     group1          69             90             88               P   
2     group1          90             95             93               P   
3     group1          47             57             44               F   
4     group1          76             78             75               P   
..       ...         ...            ...            ...             ...   
995  group10          88             99             95               P   
996  group10          62             55             55               P   
997  group10          59             71             65               P   
998  group10          68             78             77               P   
999  group10          77             86             86               P   

    Reading_PassStatus Writing_PassStatus  
0                    P                  P  
1                    P                  P  
2                    P                  P  
3                    P                  F  
4                    P                  P  
..                 ...                ...  
995                  P                  P  
996                  P                  P  
997                  P                  P  
998                  P                  P  
999                  P                  P  

[1000 rows x 12 columns]

#### Задание 15. Сколько студентов успешно сдали все экзамены?

Создайте столбец OverAll_PassStatus и запишите в него для каждого студента 'F', если студент не сдал хотя бы один из трех экзаменов, а иначе 'P'.

Посчитайте количество студентов, которые сдали все экзамены.

In [75]:
df["OverAll_PassStatus"] = "F"
df.loc[(df["Math_PassStatus"] == "P") & (df["Reading_PassStatus"] == "P") & (df["Writing_PassStatus"] == "P"), "OverAll_PassStatus"] = "P"

In [76]:
df[df["OverAll_PassStatus"] == "P"]["OverAll_PassStatus"].value_counts()

P    812
Name: OverAll_PassStatus, dtype: int64

Следовательно, 812 студентов сдали все экзамены.

#### Задание 16. Переведем баллы в оценки

### Система перевода баллов в оценки
####      90 и больше = A
####      80-89 = B
####      70-79 = C
####      60-69 = D
####      50-59 = E
####    меньше 50 = F (Fail)

Создайте вспомогательную функцию, которая будет по среднему баллу за три экзамена выставлять оценку студенту по данным выше критериям.

Создайте столбец Grade и запишите в него оценку каждого студента.

Выведите количество студентов, получивших каждую из оценок.

In [111]:
def GetGrade(average_mark):
    if average_mark >= 90:
        return "A"
    elif average_mark >= 80:
        return "B"
    elif average_mark >= 70:
        return "C"
    elif average_mark >= 60:
        return "D"
    elif average_mark >= 50:
        return "E"
    else:
        return "F"

In [112]:
df["Grade"] = ((df["math score"]+df["reading score"]+df["writing score"])/3).apply(lambda x: GetGrade(x))

In [113]:
df["Grade"].value_counts()

C    261
D    256
E    182
B    146
F    103
A     52
Name: Grade, dtype: int64